README.md
Fast Explicit Diffusion (FED).

References:

* Fast Explicit Diffusion (FED) and Fast-Jacobi (FJ):  
  <http://www.mia.uni-saarland.de/Research/SC_FED.shtml>

* S. Grewenig, J. Weickert, C. Schroers, A. Bruhn:  
 "Cyclic Schemes for PDE-Based Image Analysis"  
  Technical Report No. 327, Department of Mathematics, Saarland University, Saarbrucken, Germany, March 2013.

* S. Grewenig, J. Weickert, A. Bruhn:  
  "From box filtering to fast explicit diffusion"  
  In M. Goesele, S. Roth, A. Kuijper, B. Schiele, K. Schindler (Eds.): Pattern Recognition.  
  Lecture Notes in Computer Science, Vol. 6376, 533-542, Springer, Berlin, 2010.

CHANGELOG.md

#### 0.3.0 - Jan 25, 2014

  * Updated README.md, LICENSE.txt and CHANGELOG.md 

#### 0.2.0 - Nov 08, 2013

  * Ready for Dart 1.0.0!

#### 0.1.0 - Oct 12, 2013

  * First Dart pub package release


test/src/test_fed.dart
/*
  Copyright (c) 2013 Juan Mellado

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/

part of fed_test;

_unordererEquals(int n, double tauMax) {
  var tau = tauBySteps(n, tauMax, ReorderingMode.NONE);
  var leja = tauBySteps(n, tauMax, ReorderingMode.LEJA);

  expect(leja, isNot(orderedEquals(tau)));
  expect(leja, unorderedEquals(tau));
}

abstract class TestFed {

  static void run() {

    group("fed", () {

      test("tauBySteps - 10", () {
        var tau = tauBySteps(10, 0.5, ReorderingMode.NONE);
        expect(tau, hasLength(10));
        expect(tau[0], closeTo(0.251404, 0.000001));
        expect(tau[1], closeTo(0.263024, 0.000001));
        expect(tau[2], closeTo(0.288508, 0.000001));
        expect(tau[7], closeTo(1.33, 0.01));
        expect(tau[8], closeTo(2.88, 0.01));
        expect(tau[9], closeTo(11.25, 0.01));
        expect(tau.reduce((v, e) => v + e), closeTo(18.33, 0.01));
      });

      test("tauBySteps - 25", () {
        var tau = tauBySteps(25, 0.5, ReorderingMode.NONE);
        expect(tau, hasLength(25));
        expect(tau[0], closeTo(0.250237, 0.000001));
        expect(tau[1], closeTo(0.252147, 0.000001));
        expect(tau[2], closeTo(0.256024, 0.000001));
        expect(tau[22], closeTo(7.40, 0.01));
        expect(tau[23], closeTo(16.55, 0.01));
        expect(tau[24], closeTo(65.97, 0.01));
        expect(tau.reduce((v, e) => v + e), closeTo(108.33, 0.01));
      });

      test("tauBySteps - 50", () {
        var tau = tauBySteps(50, 0.5, ReorderingMode.NONE);
        expect(tau, hasLength(50));
        expect(tau[0], closeTo(0.250060, 0.000001));
        expect(tau[1], closeTo(0.250545, 0.000001));
        expect(tau[2], closeTo(0.251518, 0.000001));
        expect(tau[47], closeTo(28.79, 0.01));
        expect(tau[48], closeTo(64.68, 0.01));
        expect(tau[49], closeTo(258.48, 0.01));
        expect(tau.reduce((v, e) => v + e), closeTo(425.00, 0.01));
      });

      test("tauBySteps - 100", () {
        var tau = tauBySteps(100, 0.5, ReorderingMode.NONE);
        expect(tau, hasLength(100));
        expect(tau[0], closeTo(0.250015, 0.000001));
        expect(tau[1], closeTo(0.250137, 0.000001));
        expect(tau[2], closeTo(0.250382, 0.000001));
        expect(tau[97], closeTo(113.79, 0.01));
        expect(tau[98], closeTo(255.93, 0.01));
        expect(tau[99], closeTo(1023.45, 0.01));
        expect(tau.reduce((v, e) => v + e), closeTo(1683.33, 0.01));
      });

      test("tauBySteps - 250", () {
        var tau = tauBySteps(250, 0.5, ReorderingMode.NONE);
        expect(tau, hasLength(250));
        expect(tau[0], closeTo(0.250002, 0.000001));
        expect(tau[1], closeTo(0.250022, 0.000001));
        expect(tau[2], closeTo(0.250061, 0.000001));
        expect(tau[247], closeTo(706.52, 0.01));
        expect(tau[248], closeTo(1589.57, 0.01));
        expect(tau[249], closeTo(6358.01, 0.01));
        expect(tau.reduce((v, e) => v + e), closeTo(10458.33, 0.01));
      });

      test("tauBySteps - 500", () {
        var tau = tauBySteps(500, 0.5, ReorderingMode.NONE);
        expect(tau, hasLength(500));
        expect(tau[0], closeTo(0.250001, 0.000001));
        expect(tau[1], closeTo(0.250006, 0.000001));
        expect(tau[2], closeTo(0.250015, 0.000001));
        expect(tau[497], closeTo(2820.19, 0.01));
        expect(tau[498], closeTo(6345.33, 0.01));
        expect(tau[499], closeTo(25381.06, 0.01));
        expect(tau.reduce((v, e) => v + e), closeTo(41750.00, 0.01));
      });

      test("tauBySteps - 1000", () {
        var tau = tauBySteps(1000, 0.5, ReorderingMode.NONE);
        expect(tau, hasLength(1000));
        expect(tau[0], closeTo(0.250000, 0.000001));
        expect(tau[1], closeTo(0.250001, 0.000001));
        expect(tau[2], closeTo(0.250004, 0.000001));
        expect(tau[997], closeTo(11269.25, 0.01));
        expect(tau[998], closeTo(25355.72, 0.01));
        expect(tau[999], closeTo(101422.61, 0.01));
        expect(tau.reduce((v, e) => v + e), closeTo(166833.33, 0.01));
      });

      test("tauBySteps - Leja", () {
        _unordererEquals(10, 0.5);
        _unordererEquals(25, 0.5);
        _unordererEquals(50, 0.5);
        _unordererEquals(100, 0.5);
        _unordererEquals(250, 0.5);
        _unordererEquals(500, 0.5);
        _unordererEquals(1000, 0.5);
      });

      test("maxCycleTimeBySteps", () {
        expect(maxCycleTimeBySteps(10, 0.5), closeTo(18.33, 0.01));
        expect(maxCycleTimeBySteps(25, 0.5), closeTo(108.33, 0.01));
        expect(maxCycleTimeBySteps(50, 0.5), closeTo(425.00, 0.01));
        expect(maxCycleTimeBySteps(100, 0.5), closeTo(1683.33, 0.01));
        expect(maxCycleTimeBySteps(250, 0.5), closeTo(10458.33, 0.01));
        expect(maxCycleTimeBySteps(500, 0.5), closeTo(41750.00, 0.01));
        expect(maxCycleTimeBySteps(1000, 0.5), closeTo(166833.33, 0.01));
      });

      test("maxProcessTimeBySteps", () {
        expect(maxProcessTimeBySteps(10, 1, 0.5), closeTo(18.33, 0.01));
        expect(maxProcessTimeBySteps(25, 1, 0.5), closeTo(108.33, 0.01));
        expect(maxProcessTimeBySteps(50, 1, 0.5), closeTo(425.00, 0.01));
        expect(maxProcessTimeBySteps(100, 1, 0.5), closeTo(1683.33, 0.01));
        expect(maxProcessTimeBySteps(250, 1, 0.5), closeTo(10458.33, 0.01));
        expect(maxProcessTimeBySteps(500, 1, 0.5), closeTo(41750.00, 0.01));
        expect(maxProcessTimeBySteps(1000, 1, 0.5), closeTo(166833.33, 0.01));
      });

    });
  }
}


trunk/test/fed_test.dart
/*
  Copyright (c) 2013 Juan Mellado

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/

library fed_test;

import "package:unittest/unittest.dart";

import "package:fed/src/fed_util.dart";
import "package:fed/fed.dart";

part "src/test_fed_util.dart";
part "src/test_fed.dart";

// Test all.
void main() {
  TestFedUtil.run();
  TestFed.run();
}


test/src/test_fed_util.dart
/*
  Copyright (c) 2013 Juan Mellado

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/

part of fed_test;

abstract class TestFedUtil {

  static void run() {

    group("fed_util", () {

      test("nextPrime", () {
        expect(nextPrime(-2), equals(2));
        expect(nextPrime(-1), equals(2));

        expect(nextPrime(0), equals(2));
        expect(nextPrime(1), equals(2));
        expect(nextPrime(2), equals(3));
        expect(nextPrime(3), equals(5));
        expect(nextPrime(4), equals(5));
        expect(nextPrime(5), equals(7));
        expect(nextPrime(6), equals(7));
        expect(nextPrime(7), equals(11));
        expect(nextPrime(8), equals(11));
        expect(nextPrime(9), equals(11));
        expect(nextPrime(10), equals(11));
        expect(nextPrime(11), equals(13));
        expect(nextPrime(12), equals(13));
        expect(nextPrime(13), equals(17));

        expect(nextPrime(114), equals(127));
        expect(nextPrime(991), equals(997));
      });

      test("isPrime", () {
        expect(isPrime(-2), isFalse);
        expect(isPrime(-1), isFalse);

        expect(isPrime(0), isFalse);
        expect(isPrime(1), isFalse);
        expect(isPrime(2), isTrue);
        expect(isPrime(3), isTrue);
        expect(isPrime(4), isFalse);
        expect(isPrime(5), isTrue);
        expect(isPrime(6), isFalse);
        expect(isPrime(7), isTrue);
        expect(isPrime(8), isFalse);
        expect(isPrime(9), isFalse);
        expect(isPrime(10), isFalse);
        expect(isPrime(11), isTrue);
        expect(isPrime(12), isFalse);
        expect(isPrime(13), isTrue);

        expect(isPrime(121), isFalse);
        expect(isPrime(997), isTrue);
      });

    });
  }
}


lib/src/fed_util.dart
/*
  Copyright (c) 2013 Juan Mellado

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/

/*
This code is derived from FED/FJ library.

References:
- Fast Explicit Diffusion (FED) and Fast-Jacobi (FJ):
  http://www.mia.uni-saarland.de/Research/SC_FED.shtml

- S. Grewenig, J. Weickert, C. Schroers, A. Bruhn:
  "Cyclic Schemes for PDE-Based Image Analysis"
  Technical Report No. 327, Department of Mathematics, Saarland University, Saarbrücken, Germany, March 2013.

- S. Grewenig, J. Weickert, A. Bruhn:
  "From box filtering to fast explicit diffusion"
  In M. Goesele, S. Roth, A. Kuijper, B. Schiele, K. Schindler (Eds.): Pattern Recognition.
  Lecture Notes in Computer Science, Vol. 6376, 533-542, Springer, Berlin, 2010.
*/

library fed_util;

import "dart:math" as math;

/**
 * Returns the next prime number greater than [number].
 */
int nextPrime(int number) {
  var prime = number + 1;
  while (!isPrime(prime)) {
    ++ prime;
  }
  return prime;
}

/**
 * Checks if the given [number] is prime.
 */
bool isPrime(int number) {
  if (number <= 1) {
    return false;
  }
  if ((number <= 3) || (number == 5) || (number == 7)) {
    return true;
  }
  if (((number % 2) == 0) || ((number % 3) == 0) || ((number % 5) == 0) || ((number % 7) == 0)) {
    return false;
  }

  var divisor = 11;
  var upperLimit = math.sqrt(number + 1).toInt();

  while(divisor <= upperLimit) {
    if ((number % divisor) == 0) {
      return false;
    }
    divisor += 2;
  }

  return true;
}


trunk/lib/fed.dart
/*
  Copyright (c) 2013 Juan Mellado

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/

/*
This code is derived from FED/FJ library.

References:
- Fast Explicit Diffusion (FED) and Fast-Jacobi (FJ):
  http://www.mia.uni-saarland.de/Research/SC_FED.shtml

- S. Grewenig, J. Weickert, C. Schroers, A. Bruhn:
  "Cyclic Schemes for PDE-Based Image Analysis"
  Technical Report No. 327, Department of Mathematics, Saarland University, Saarbrücken, Germany, March 2013.

- S. Grewenig, J. Weickert, A. Bruhn:
  "From box filtering to fast explicit diffusion"
  In M. Goesele, S. Roth, A. Kuijper, B. Schiele, K. Schindler (Eds.): Pattern Recognition.
  Lecture Notes in Computer Science, Vol. 6376, 533-542, Springer, Berlin, 2010.
*/

library fed;

import "dart:math" as math;

import "package:fed/src/fed_util.dart";

part "src/fed_kappa.dart";

/**
 * Allocates an array of [n] time steps and fills it with FED time step sizes,
 * such that the maximal stopping time for this cycle is obtained. [tauMax] is
 * the stability limit. Use [reordering] if time steps should be reordered
 * using kappa cycles.
 *
 * See [ReorderingMode]
 */
List<double> tauBySteps(int n, double tauMax, KappaLookup reordering) {
  return _tau(n, 1.0, tauMax, reordering);
}

/**
 * Allocates an array of the least number of time steps such that a certain
 * stopping time per cycle [t] can be obtained and fills it with the respective
 * FED time step sizes. [tauMax] is the stability limit. Use [reordering] if
 * time steps should be reordered using kappa cycles.
 *
 * See [ReorderingMode]
 */
List<double> tauByCycleTime(double t, double tauMax, KappaLookup reordering) {
  // Number of time steps.
  var n = ((math.sqrt(((3 * t) / tauMax) + 0.25) - 0.5 - 1.0e-8)).ceil();

  // Ratio of t we search to maximal t.
  var scale = (3 * t) / (tauMax * (n * (n + 1)));

  return _tau(n, scale, tauMax, reordering);
}

/**
 * Allocates an array of the least number of time steps such that a certain
 * stopping time [t] for the whole process consisting of [m] cycles can be
 * obtained, and fills it with the respective FED time step sizes for one cycle.
 * [tauMax] is the stability limit. Use [reordering] if time steps should be
 * reordered using kappa cycles.
 *
 * See [ReorderingMode]
 */
List<double> tauByProcessTime(double t, int m, double tauMax, KappaLookup reordering) {
  return tauByCycleTime(t / m, tauMax, reordering);
}

/**
 * Computes the maximal cycle time that can be obtained using a certain
 * number of steps [n]. This corresponds to the cycle time that arises from a
 * tau array which has been created using [:tauBySteps(n, tauMax, reordering):].
 * [tauMax] is the stability limit.
 */
double maxCycleTimeBySteps(int n, double tauMax) {
  return (tauMax * (n * (n + 1))) / 3;
}

/**
 * Computes the maximal process time that can be obtained using a certain
 * number of steps [n] for the whole process consisting of [m] cycles. This
 * corresponds to the cycle time that arises from a tau array which has been
 * created using [:tauBySteps(n, tauMax, reordering):]. [tauMax] is the stability
 * limit.
 */
double maxProcessTimeBySteps(int n, int m, double tauMax) {
  return (tauMax * (n * (n + 1) * m)) / 3;
}

/**
 * Allocates an array of [n] relaxation parameters and fills it with the FED
 * based parameters for Fast-Jacobi. [omegaMax] is the stability limit. Use
 * [reordering] if time steps should be reordered using kappa cycles.
 *
 * See [ReorderingMode]
 */
List<double> relaxParams(int n, double omegaMax, KappaLookup reordering) {
  return _tau(n, 1.0, omegaMax, reordering);
}

/**
 * Allocates an array of [n] time steps and fills it with FED time step sizes.
 */
List<double> _tau(int n, double scale, double tauMax, KappaLookup reordering) {
  var tau = new List<double>(n);

  var c = math.PI / ((4 * n) + 2);
  var d = (scale * tauMax) / 2;
  var c2 = 2 * c;

  if (reordering == ReorderingMode.NONE) {
    for (var i = 0; i < n; ++ i) {
      var h = math.cos(c + (c2 * i));
      tau[i] = d / (h * h);
    }
  } else {
    var kappa = reordering(n);
    var prime = nextPrime(n);
    var k = 0;
    var index;

    for (var i = 0; i < n; ++ i) {
      while ((index = (k += kappa) % prime) > n)
      ;

      var h = math.cos(c + (c2 * -- index));
      tau[i] = d / (h * h);
    }
  }

  return tau;
}
